### PR TITLE
fix: resolve 5 critical deployment issues blocking CI/CD workflows (#481)

### LINKED ISSUE #481: [Framework] 5 Critical Deployment Issues Requiring Framework Fixes
# Framework Deployment Issues: 5 Critical Fixes Needed

**Summary**: Multiple deployment issues discovered during frontify-frigg integration that require framework-level fixes. These issues force every Frigg user to implement CI/CD workarounds instead of being handled by the framework itself.

## Priority Order
1. **Issue 3** (esbuild directories) - Blocks all CI/CD deployments ⚠️ CRITICAL
2. **Issue 5** (AWS discovery opt-out) - UX improvement for restrictive credentials
3. **Issue 2** (Prisma cleanup) - Blocks deployments randomly
4. **Issue 1** (osls dependency) - Easy one-line fix
5. **Issue 4** (plugin conflicts) - Cleanup task

---

### Problem
`@friggframework/frigg-cli` spawns `osls` (OSS Serverless) as a subprocess but doesn't declare it as a dependency.

### Impact
- Users must install `osls` globally: `npm install -g osls`
- CI/CD workflows require extra installation step
- Violates npm dependency management best practices
- Error: `spawn osls ENOENT`

### Evidence
```javascript
// packages/frigg-cli/deploy-command/index.js
const command = 'osls';  // Subprocess call without dependency

// packages/frigg-cli/build-command/index.js
const command = 'osls';  // Same issue
```

### Fix Required
```json
// packages/frigg-cli/package.json
{
  "dependencies": {
    "osls": "^3.40.1"
  }
}
```

### Current Workaround
```yaml
# Every user's .github/workflows/deploy.yml
- name: Install osls globally
  run: npm install -g osls
```

**Priority**: Medium
**Files**: `packages/frigg-cli/deploy-command/index.js`, `packages/frigg-cli/build-command/index.js`, `packages/frigg-cli/package.json`

---

### Problem
Prisma Lambda layer build script doesn't clean up stale artifacts, causing ENOTEMPTY errors when previous builds failed/interrupted.

### Impact
- Deployment fails: `ENOTEMPTY: directory not empty`
- Corrupt layer directories block subsequent deployments
- Requires manual cleanup
- Cryptic error messages confuse users

### Evidence
```
npm error syscall rmdir
npm error path /backend/layers/prisma/nodejs/node_modules/@prisma/client/runtime
npm error errno -39
npm error ENOTEMPTY: directory not empty, rmdir '...'
```

### Root Cause
`buildPrismaLayer()` creates directories but never cleans existing ones before starting.

### Fix Required (Recommended)
```javascript
// packages/devtools/infrastructure/scripts/build-prisma-layer.js
async function buildPrismaLayer(config) {
    const { layerPath, targetEnv } = config;

    // Add cleanup at the beginning
    if (fs.existsSync(layerPath)) {
        console.log(`Cleaning existing layer directory: ${layerPath}`);
        await fs.rm(layerPath, { recursive: true, force: true });
    }

    // ... rest of build logic
}
```

### Additional Improvements
1. Add retry logic for npm install failures
2. Better error messages explaining the issue
3. Add `--clean` flag to frigg deploy command
4. Implement proper rollback on build failure

### Current Workaround
```yaml
# Every user's .github/workflows/deploy.yml
- name: Clean Prisma layer directory
  run: rm -rf layers/prisma
```

**Priority**: High
**Files**: `packages/devtools/infrastructure/scripts/build-prisma-layer.js`, `packages/devtools/infrastructure/domains/shared/utilities/prisma-layer-manager.js`

---

### Problem
`serverless-esbuild` expects `.esbuild/.serverless` directory to exist, but Frigg doesn't create it.

**IMPORTANT**: Since Frigg's infrastructure builder adds `serverless-esbuild` to the plugins list, **Frigg should handle creating the required directories**. This is Frigg's responsibility, not the user's.

### Impact
- Deployment fails: `ENOENT: no such file or directory, lstat '.esbuild/.serverless'`
- **Blocks ALL CI/CD deployments**
- Not reproducible locally (directory persists after first run)
- Forces every Frigg user to add workarounds

### Evidence
```
Error: ENOENT: no such file or directory, lstat '/home/runner/work/frigg-integrations/frigg-integrations/backend/.esbuild/.serverless'
× Stack create-frigg-app-production failed to deploy (128s)
```

### Root Cause
1. Frigg infrastructure builder adds `serverless-esbuild` to plugins
2. Plugin expects `.esbuild/.serverless` directory for build artifacts
3. **Frigg adds the plugin but doesn't handle its requirements** ← Core issue
4. CI/CD environments start clean, directory never exists

### Fix Required (BEST Option)
```javascript
// packages/devtools/serverless-plugin/index.js
class FriggServerlessPlugin {
    constructor(serverless, options) {
        this.serverless = serverless;
        this.options = options;

        this.hooks = {
            'before:package:initialize': this.ensureBuildDirectories.bind(this),
        };
    }

    async ensureBuildDirectories() {
        const fs = require('fs').promises;

        const buildDirs = ['.esbuild', '.esbuild/.serverless'];

        for (const dir of buildDirs) {
            await fs.mkdir(dir, { recursive: true });
        }

        this.serverless.cli.log('✓ Created build directories for serverless-esbuild');
    }
}
```

**Why this is best**:
- Frigg serverless plugin already runs during deployment
- Hooks into `before:package:initialize` lifecycle event
- Centralized location for Frigg-specific setup
- Won't run if serverless-esbuild isn't in plugins
- Follows separation of concerns

### Alternative Options
- **Option 2**: Add to `packages/frigg-cli/deploy-command/index.js` before spawning osls
- **Option 3**: Add to `packages/devtools/infrastructure/infrastructure-composer.js`

### Current Workaround
```yaml
# Every user's .github/workflows/deploy.yml
- name: Create esbuild directories
  run: mkdir -p .esbuild/.serverless
```

**Priority**: CRITICAL
**Files**: `packages/devtools/serverless-plugin/index.js`, `packages/devtools/infrastructure/domains/shared/utilities/base-definition-factory.js`

---

### Problem
Frigg infrastructure builder adds `serverless-esbuild`, but legacy apps have `serverless-jetpack`. Both handle packaging and may conflict.

### Impact
- Unclear which plugin packages Lambda functions
- Potential configuration conflicts
- Unnecessary dependencies and build time
- No migration guide

### Evidence
```json
// User's backend/package.json
{
  "devDependencies": {
    "serverless-esbuild": "^1.55.1",  // Added by Frigg
    "serverless-jetpack": "0.11.2"     // Legacy
  }
}
```

### Fix Required (Recommended)
**Choose serverless-esbuild** and provide migration path:

1. Update infrastructure builder to check for existing plugins
2. Warn if both are present
3. Document migration from jetpack to esbuild
4. Update create-frigg-app templates

**Alternative**: Smart detection:
```javascript
// base-definition-factory.js
function getPackagingPlugin(packageJson) {
    const hasJetpack = packageJson.devDependencies?.['serverless-jetpack'];
    const hasEsbuild = packageJson.devDependencies?.['serverless-esbuild'];

    if (hasJetpack && hasEsbuild) {
        console.warn('Both plugins found. Using serverless-esbuild.');
    }

    return hasEsbuild ? 'serverless-esbuild' :
           hasJetpack ? 'serverless-jetpack' :
           'serverless-esbuild'; // default
}
```

**Recommendation**: serverless-esbuild
- More modern
- Better TypeScript support
- Faster builds
- Active development

**Priority**: Medium
**Files**: `packages/devtools/infrastructure/domains/shared/utilities/base-definition-factory.js`

---

### Problem
AWS resource discovery fails silently when IAM credentials lack permissions. No way to explicitly opt-out for restrictive deployment credentials.

### Impact
- Users see discovery errors but deployment continues
- Unclear if deployment will work without discovered resources
- No explicit discovery toggle
- Forces users to grant broad permissions or deal with confusing errors

### Evidence
```
Error discovering AWS resources: User: arn:aws:iam::123456789012:user/github-deploy
is not authorized to perform: ec2:DescribeVpcs on resource: arn:aws:ec2:eu-central-1:123456789012:vpc/*

Warning: AWS discovery failed, continuing with deployment...
```

### Root Cause
From `packages/devtools/infrastructure/CLAUDE.md`:
```javascript
const shouldRunDiscovery = (AppDefinition) => {
    return (
        AppDefinition.vpc?.enable === true ||
        AppDefinition.encryption?.useDefaultKMSForFieldLevelEncryption === true ||
        AppDefinition.ssm?.enable === true
    );
};
```

Discovery runs automatically but:
1. No way to disable it
2. Failures logged as warnings, deployment continues
3. No validation that required permissions exist
4. No manual override option

### Fix Required (Recommended)
```javascript
// AppDefinition schema - add discovery toggle
const appDefinition = {
    aws: {
        discovery: {
            enabled: true,              // Explicit opt-in/opt-out
            failOnError: false,         // Fail deployment if discovery fails
            requiredResources: []       // Specify what must be discovered
        }
    },
    vpc: {
        enable: true,
        vpcId: 'vpc-12345',            // Manual override if discovery disabled
        subnetIds: ['subnet-1', 'subnet-2'],
        securityGroupIds: ['sg-123']
    }
};
```

```javascript
// packages/devtools/infrastructure/aws-discovery.js - improve error handling
async function discoverResources(appDefinition) {
    if (appDefinition.aws?.discovery?.enabled === false) {
        console.log('ℹ️  AWS discovery disabled, using manual configuration');
        return null;
    }

    try {
        const resources = await performDiscovery();
        return resources;
    } catch (error) {
        const message = `
❌ AWS Discovery Failed

The deployment credential doesn't have permissions for resource discovery.

Options:
1. Grant additional IAM permissions (see docs/IAM-POLICY-TEMPLATES.md)
2. Disable discovery and provide resources manually:

   aws: { discovery: { enabled: false } }
   vpc: { vpcId: 'vpc-xxx', subnetIds: [...] }

3. Continue without VPC/KMS features

Error: ${error.message}
        `;

        if (appDefinition.aws?.discovery?.failOnError === true) {
            throw new Error(message);
        } else {
            console.warn(message);
            return null;
        }
    }
}
```

### Additional Improvements
1. Add `--skip-discovery` flag to frigg deploy command
2. Validate manual configuration when discovery disabled
3. Document minimal IAM policy (deployment-only, no discovery)
4. Add IAM permission checker before running discovery
5. Better error messages per discovery step

### Required IAM Permissions

**Minimal (deployment-only)**:
- `cloudformation:*`
- `lambda:*`
- `apigateway:*`
- `iam:CreateRole`, `iam:AttachRolePolicy`
- `logs:CreateLogGroup`, `logs:PutRetentionPolicy`

**Optional (for discovery)**:
- `ec2:DescribeVpcs`, `ec2:DescribeSubnets`, `ec2:DescribeSecurityGroups`
- `kms:ListKeys`, `kms:DescribeKey`, `kms:ListAliases`
- `ssm:GetParameter`, `ssm:PutParameter`

**Priority**: Medium
**Files**: `packages/devtools/infrastructure/aws-discovery.js`, `packages/devtools/infrastructure/build-time-discovery.js`

---

## Summary

All 5 issues force users to implement workarounds in their CI/CD pipelines. **These should be fixed in the framework.**

### Why This Matters
- Every Frigg user encounters these issues
- Workarounds are duplicated across all projects
- Poor developer experience
- Violates framework's "spin up integrations in minutes" promise

### Recommended Implementation Order
1. **Issue 3** - Highest impact, quick serverless plugin fix
2. **Issue 5** - UX improvement, quick config addition
3. **Issue 2** - Random failures, moderate effort
4. **Issue 1** - One-line package.json fix
5. **Issue 4** - Cleanup task, document migration

### Files Affected
- `packages/frigg-cli/package.json`
- `packages/frigg-cli/deploy-command/index.js`
- `packages/devtools/serverless-plugin/index.js`
- `packages/devtools/infrastructure/scripts/build-prisma-layer.js`
- `packages/devtools/infrastructure/aws-discovery.js`
- `packages/devtools/infrastructure/build-time-discovery.js`
- `packages/devtools/infrastructure/domains/shared/utilities/base-definition-factory.js`

### PR DESCRIPTION
**Pull request by seanspeaks**
# Fix Critical Deployment Issues from GitHub Issue #481

This PR completely resolves all 5 deployment issues identified in #481 that were forcing users to implement custom CI/CD workarounds instead of having the framework handle them natively.

### 🔴 Issue 3: Missing esbuild Directories (CRITICAL) - **Enhanced with Constructor Fix**

**Problem:** serverless-esbuild plugin expected `.esbuild/.serverless` directories that didn't exist, blocking ALL CI/CD deployments in clean environments.

**Critical Discovery:** Initial hook-based fix had a race condition - serverless-esbuild could initialize before Frigg plugin hooks ran.

**Solution:** Moved directory creation to plugin **constructor** (synchronous, guaranteed first):

```javascript
constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;
    
    // Create directory IMMEDIATELY, before ANY hooks run
    const fs = require('fs');
    const esbuildDir = path.join(/* ... */, '.esbuild', '.serverless');
    fs.mkdirSync(esbuildDir, { recursive: true });
    
    this.hooks = { /* ... */ }; // Hooks registered AFTER directory exists
}
```

**Why Constructor?**
- ✅ Runs synchronously before hooks are registered
- ✅ Guaranteed to execute before serverless-esbuild loads
- ✅ No race conditions possible
- ✅ Directory exists immediately when needed

---

### 🟡 Issue 5: AWS Discovery Silent Failures - **Enhanced with AppDefinition Control**

**Problem:** Discovery failures were silent with no way to explicitly disable for restrictive IAM credentials.

**Solution:** Added **three-tier discovery control** with `failOnError` flag:

**1. AppDefinition-level configuration (Priority 1):**
```javascript
module.exports = {
    aws: {
        discovery: {
            enabled: false,      // Explicit disable for restrictive IAM
            failOnError: true    // Fail deployment on discovery errors
        }
    }
};
```

**2. Environment variable (Priority 2 - backward compatible):**
```bash
export FRIGG_SKIP_AWS_DISCOVERY=true
```

**3. Auto-detection (Priority 3):**
Automatically runs when VPC, KMS, SSM, or PostgreSQL features are enabled.

**Use Cases:**
- **Restrictive IAM in CI/CD:** `aws.discovery.enabled: false`
- **Strict production:** `aws.discovery.failOnError: true` (fail if discovery fails)
- **Graceful dev:** `aws.discovery.failOnError: false` (default - continue on failure)

---

### ✅ Issue 1: Missing osls Dependency

**Problem:** frigg-cli spawned osls subprocess without declaring it as a dependency.

**Solution:** Added `osls: "^3.40.1"` to `packages/frigg-cli/package.json` dependencies.

**Impact:** No more manual `npm install -g osls` required in CI/CD pipelines.

---

### ✅ Issue 2: Prisma Layer Build Cleanup

**Status:** Already fixed in `next` branch.

**Implementation:** `cleanLayerDirectory()` function in `build-prisma-layer.js:166-175` automatically removes stale artifacts before each build.

---

### ✅ Issue 4: Conflicting Packaging Plugins

**Problem:** Both serverless-esbuild and serverless-jetpack could coexist, causing unclear build behavior.

**Solution:** Created `plugin-validator.js` with:
- **Auto-detection** of plugin conflicts
- **Auto-fix** removes serverless-jetpack when esbuild is present
- **Migration warnings** for legacy configurations with clear recommendations

**Integration:**
```javascript
// In infrastructure-composer.js
const pluginValidation = validateAndCleanPlugins(definition.plugins, {
    autoFix: true,
    silent: false,
});

if (pluginValidation.modified) {
    definition.plugins = pluginValidation.plugins;
    console.log('   ✓ Plugin configuration auto-fixed');
}
```

---

## 📊 Summary Table

| Issue | Status | Priority | Implementation |
|-------|--------|----------|---------------|
| #1 osls dependency | ✅ Fixed | Medium | Added to package.json |
| #2 Prisma cleanup | ✅ Pre-existing | High | cleanLayerDirectory() |
| #3 esbuild directories | ✅ **Enhanced** | **CRITICAL** | **Constructor-based (timing fix)** |
| #4 Plugin conflicts | ✅ Fixed | Medium | Auto-detection + validation |
| #5 Discovery failures | ✅ **Enhanced** | Medium | **AppDefinition + failOnError** |

---

### New Files
- ✨ `packages/devtools/infrastructure/domains/shared/validation/plugin-validator.js` - Plugin conflict detection
- ✨ `packages/devtools/infrastructure/domains/shared/validation/plugin-validator.test.js` - 16 test cases
- ✨ `packages/devtools/infrastructure/domains/shared/resource-discovery.enhanced.test.js` - Discovery config tests
- ✨ `packages/frigg-cli/__tests__/unit/dependencies.test.js` - Dependency validation tests
- ✨ `docs/reference/deployment-fixes.md` - Complete documentation with examples

### Modified Files
- 🔧 `packages/frigg-cli/package.json` - Added osls dependency
- 🔧 `packages/serverless-plugin/index.js` - **Constructor-based directory creation**
- 🔧 `packages/devtools/infrastructure/infrastructure-composer.js` - Plugin validation integration
- 🔧 `packages/devtools/infrastructure/domains/shared/resource-discovery.js` - AppDefinition control

---

### Architecture Compliance
✅ Follows **Hexagonal Architecture** (Ports & Adapters):
- **Domain Layer:** Plugin validator, resource discovery logic
- **Application Layer:** Infrastructure composer orchestration
- **Infrastructure Layer:** Plugin constructor (early initialization), package dependencies
- **Clear separation** of configuration from implementation

### Testing Approach
✅ Follows **TDD best practices**:
- **25+ test cases** covering all scenarios
- Unit tests for plugin conflict detection
- Integration tests for auto-fix behavior
- Dependency validation tests
- Edge case coverage (empty arrays, undefined values, timing)

---

## 🔄 Backward Compatibility

✅ **All existing configurations continue to work:**
- Environment variable `FRIGG_SKIP_AWS_DISCOVERY=true` still supported
- Auto-detection based on features still works
- Graceful error handling is default (`failOnError` defaults to false)
- No breaking changes to existing APIs

---

## 📚 Documentation

Comprehensive documentation added in `docs/reference/deployment-fixes.md`:
- Detailed problem analysis for each issue
- Solution explanations with code examples
- Migration guide for existing projects
- Usage scenarios (CI/CD, production, dev environments)
- Priority matrices for discovery control
- Evidence of timing issues with logs

---

### For Existing Projects

**1. Update dependencies:**
```bash
npm install
```

**2. If using serverless-jetpack, remove it (auto-detected with warning):**
```yaml
# serverless.yml
plugins:
  - serverless-esbuild  # Keep
  # - serverless-jetpack  # Remove (auto-fixed with warning)
```

**3. For restrictive IAM in CI/CD, use AppDefinition:**
```javascript
// index.js
module.exports = {
    name: 'my-app',
    aws: { discovery: { enabled: false } }
};
```

### For New Projects

✅ **No action required!** All fixes work automatically when using:
```bash
npx @friggframework/frigg-cli init my-project
```

---

## ✨ Benefits

- 🎯 **No more CI/CD workarounds** - Framework handles everything
- ⚡ **No race conditions** - Constructor-based esbuild fix eliminates timing issues
- 🔧 **Configuration as code** - AppDefinition-level discovery control
- 📦 **Proper dependencies** - No manual global installations required
- 🛡️ **Production safety** - Strict mode with `failOnError: true`
- 📝 **Clear logging** - Know exactly what the framework is doing
- 🧪 **Comprehensive tests** - 25+ test cases ensure reliability

---

## 🔗 Related

- Closes #481
- Addresses all 5 deployment issues that were blocking CI/CD workflows
- Eliminates need for custom workarounds in deployment pipelines
- Restores framework's promise to "spin up integrations in minutes"

---

<details>
  <summary>📦 Published PR as canary version: <code>2.0.0--canary.482.bda352c.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @friggframework/core@2.0.0--canary.482.bda352c.0
  npm install @friggframework/devtools@2.0.0--canary.482.bda352c.0
  npm install @friggframework/eslint-config@2.0.0--canary.482.bda352c.0
  npm install @friggframework/prettier-config@2.0.0--canary.482.bda352c.0
  npm install @friggframework/schemas@2.0.0--canary.482.bda352c.0
  npm install @friggframework/serverless-plugin@2.0.0--canary.482.bda352c.0
  npm install @friggframework/test@2.0.0--canary.482.bda352c.0
  npm install @friggframework/ui@2.0.0--canary.482.bda352c.0
  # or 
  yarn add @friggframework/core@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/devtools@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/eslint-config@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/prettier-config@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/schemas@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/serverless-plugin@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/test@2.0.0--canary.482.bda352c.0
  yarn add @friggframework/ui@2.0.0--canary.482.bda352c.0
  ```
</details>

### CODE REVIEW

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.48`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.48) :rocket:

File: packages/devtools/frigg-cli/deploy-command/index.js [hunk 1]
```diff
@@ -113,7 +113,7 @@ function validateAndBuildEnvironment(appDefinition, options) {
 
     // Try to use the env-validator if available
     try {
-        const { validateEnvironmentVariables } = require('@friggframework/devtools/infrastructure/env-validator');
+        const { validateEnvironmentVariables } = require('../../infrastructure/env-validator');
```
Comment (d-klotz) [inline packages/devtools/frigg-cli/deploy-command/index.js:116]: Are you sure this is a good idea?

Comment (seanspeaks) [reply packages/devtools/frigg-cli/deploy-command/index.js:116]: Basically trying to release frigg-cli as its own package, while inside the devtools package, and while requiring the devtools package, was making the CD environment do a double deploy.

Ultimately because there were two installs of the devtools package (under root, and then under /frigg-cli/node_modules).

So just making the CLI part of the devtools package resolve that; which means everything is relative pathing 👍  the cli runs from here.

@d-klotz

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.49`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.49) :rocket:

**Comment by seanspeaks** [comment]
:rocket: PR was released in [`v2.0.0-next.50`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.50) :rocket:
